### PR TITLE
Fixed pandas time index duplication by rounding the floating point vector

### PR DESCRIPTION
**Pull request by choldgraf**
This should solve #1753. I think the problem was floating point precision. The time vector is stored in seconds, so when you increase these to something large by scaling, the floating point precision is slightly off. If you convert straight to int64, then you'll get some duplicated numbers. However, if you round the floating point numbers first, then the integer conversion works correctly.

### CODE REVIEW

File: mne/tests/test_epochs.py [hunk 1]
```diff
@@ -1254,5 +1254,34 @@ def test_array_epochs():
     assert_equal(epochs.drop_log[0], ['EEG 006'])
     assert_equal(len(events), len(epochs.selection))
 
+def test_dataframe_time_indexing():
```
Comment (Eric89GXL) [inline mne/tests/test_epochs.py]: two newlines between functions (PEP8)

File: mne/tests/test_epochs.py [hunk 2]
```diff
@@ -1254,5 +1254,34 @@ def test_array_epochs():
...
+    '''Make sure that conversion to dataframe
+    doesn't duplicate time indices due to rounding
+    error + floating point precision'''
```
Comment (Eric89GXL) [inline mne/tests/test_epochs.py]: By convention these are one line, my suggestion
```
"""Test that dataframe conversion doesn't duplicate time indices"""
```

File: mne/tests/test_epochs.py [hunk 3]
```diff
@@ -1254,5 +1254,34 @@ def test_array_epochs():
...
+    info = create_info(ch_names, 1000, ch_types=ch_types)
+    raw = io.RawArray(data_raw, info)
+    events = find_events(raw, 'stim')
+    epochs = Epochs(raw, events, 1, -.2, 1.5)
+
+    # Convert to dataframe
+    epochs_df = epochs.as_data_frame()
+    epochs_times = epochs_df.index.get_level_values('time')
+
+    assert epochs_times.duplicated().values.sum() == 0, 'Conversion to DataFrame duplicated timepoints'
```
Comment (Eric89GXL) [inline mne/tests/test_epochs.py]: use `assert_true(...)` or even better `assert_equal`, we use `nose` not `py.test` standards around here :)

File: mne/tests/test_epochs.py [hunk 4]
```diff
@@ -1254,5 +1254,34 @@ def test_array_epochs():
...
+    data_event[[2, 2000]] = 1  # Insert some events
+    data_raw = np.vstack([data, data_event])
+
+    # For the Info object
+    ch_types = ['eeg']*data_raw.shape[0]
+    ch_names = ['{0}'.format(i) for i in range(data_raw.shape[0])]
+    ch_names[-1] = 'stim'
+    ch_types[-1] = 'misc'
+
+
```
Comment (Eric89GXL) [inline mne/tests/test_epochs.py]: one newline instead of two

File: mne/tests/test_epochs.py [hunk 5]
```diff
@@ -1255,4 +1255,33 @@ def test_array_epochs():
...
+    ch_types = ['eeg']*data_raw.shape[0]
+    ch_names = ['{0}'.format(i) for i in range(data_raw.shape[0])]
+    ch_names[-1] = 'stim'
+    ch_types[-1] = 'misc'
+
+    # Create base MNE objects
+    info = create_info(ch_names, 1000, ch_types=ch_types)
+    raw = io.RawArray(data_raw, info)
+    events = find_events(raw, 'stim')
+    epochs = Epochs(raw, events, 1, -.2, 1.5)
```
Comment (Eric89GXL) [inline mne/tests/test_epochs.py]: Maybe this is a dumb question, but why not just use `EpochsArray` directly?

Even if you don't do that, you can avoid adding the `stim` channel by passing the `events` array directly, saves an import and use of `find_events` as well

File: mne/tests/test_epochs.py [hunk 6]
```diff
@@ -1255,4 +1255,33 @@ def test_array_epochs():
...
+    doesn't duplicate time indices due to rounding
+    error + floating point precision'''
+    # Synthesize data
+    data = np.random.randn(2, 10000)
+    data_event = np.zeros(data.shape[1])
+    data_event[[2, 2000]] = 1  # Insert some events
+    data_raw = np.vstack([data, data_event])
+
+    # For the Info object
+    ch_types = ['eeg']*data_raw.shape[0]
```
Comment (Eric89GXL) [inline mne/tests/test_epochs.py]: We follow the convention that arithmetic operators should have whitespace between them so `['eeg'] * data_raw.shape[0]` (even though `scipy` and other packages have abandoned that standard)

**Comment by Eric89GXL** [comment]
A few little nitpicks from me, but the fix looks reasonable to me. +1 for merge but I'll let @dengemann take a look and merge if he's happy since he knows more about pandas

**Comment by choldgraf** [comment]
Thanks for the comments - made a few adjustments per your thoughts.

**Comment by agramfort** [comment]
looks good but have a look at travis failures. pandas is a weak dependency

**Comment by Eric89GXL** [comment]
What my esteemed colleague is hinting at is that you look into our use of the `requires_pandas` decorator and use it for your new test.

**Comment by agramfort** [comment]
> What my esteemed colleague

:)

**Comment by Eric89GXL** [comment]
I thought Europeans were supposed to have good sarcasm detectors

**Comment by choldgraf** [comment]
Ah I see - so requires_pandas is a way of checking whether or not the packages is present, and if not skipping the test and logging it. Cool! I've added the decorator to my test.

**Comment by dengemann** [comment]
Tell me once you're done, I'll take a final look and merge then.

**Comment by agramfort** [comment]
+1 for merge.

**Comment by choldgraf** [comment]
I believe that I've addressed all comments made so far, so feel free to take a look and let me know what you think. And thanks for the input everybody.

**Comment by Eric89GXL** [comment]
AppVeyor failure is unrelated

**Comment by dengemann** [comment]
@choldgraf do we also have this problem for the other exporters (Raw, Evoked, SourceEstimate)?

**Comment by choldgraf** [comment]
Possibly - I will check them soon. If you can think of any others that need checking let me know. As an aside, before I found this I had written some of my own code to export a dataframe from some MNE objects. E.g., for the TFR Average object, I have this:

```
def convert_tfr_to_dataframe(tfr, scaling=1e3):
    df = []
    for elec, vals in zip(tfr.ch_names, tfr.data):
        ix_freq = pd.Index(tfr.freqs, name='freq')
        ix_time = pd.Index(tfr.times * scaling, name='time')
        ix_elec = pd.Index([elec]*vals.shape[0], name='elec')
        df.append(pd.DataFrame(vals, index=[ix_elec, ix_freq], columns=ix_time))
    df = pd.concat(df, axis=0)
    return df
```

Not sure how it stands in terms of efficiency and robustness and such, but I find it to be pretty straightforward for my uses. Is that something that (at least for the AverageTFR objects) would be worth adding? Or do you wish to keep "as_data_frame" consistent across modules?

**Comment by dengemann** [comment]
@choldgraf we should probably unify the code path and use a mixin pattern here. Raw, Evoked,, Epochs, TFRAverage, SourceEstimate woulde then subclass from PandasMixin specifying the as_data_frame method. Depending on the "host class", different blocks would be executed.

**Comment by choldgraf** [comment]
Would this be a very complicated change? Agree it sounds more efficient and
easier to debug. Something I could try to work on in the coming weeks if
it's worth the time, but I'm not sure how much this would require changing
tests etc to get it working.

On Thu, Jan 22, 2015 at 3:03 PM, Denis A. Engemann <notifications@github.com

**Comment by dengemann** [comment]
> Would this be a very complicated change? Agree it sounds more efficient and
easier to debug. Something I could try to work on in the coming weeks if
it's worth the time, but I'm not sure how much this would require changing
tests etc to get it working.

no change in tests. Basically copy and paste of existing code. For various mixin examples see:
https://github.com/mne-tools/mne-python/blob/master/mne/channels/channels.py#L126

**Comment by choldgraf** [comment]
OK, I can try this. Should it be a separate PR or overwrite this one?

**Comment by choldgraf** [comment]
While I'm thinking about this I wanted to bring up a question about naming. How do people feel about calling the dataframe methods "to_data_frame" rather than "as_data_frame". The reason I say this is that many objects already implement a "to_nitime" method, and it seems like both of these should have the same naming convention since they both convert the data to some other format interactively. Does that make sense?

**Comment by Eric89GXL** [comment]
+1 for increasing consistency, either as `to_datafame` or changing nitime to `as_nitime`. But I think you're right that `to_` is clearer, so I have a slight preference for that. If others concur, feel free to implement that while you're at it, deprecating the old function.

**Comment by agramfort** [comment]
to_dataframe is fine with me

+1

**Comment by choldgraf** [comment]
Where would the mixin code go? In io.base? Or create a separate dataframe module?

**Comment by agramfort** [comment]
Fine with io.base if it's not already too big a file

**Comment by choldgraf** [comment]
OK, PR for the mixin is #1760

**Comment by dengemann** [comment]
And here after #1760 for #1837

**Comment by choldgraf** [comment]
@dengemann I believe that the mixin PR would nullify this one, since it moves the definition of to_data_frame to the base module. No?

**Comment by dengemann** [comment]
yeah, probably you're right! feel free to move things there and close.

**Comment by choldgraf** [comment]
OK, I'm closing this then, all the changes should already be implemented in #1760

**Pull request closed without merging**